design.md

```diff
@@ -0,0 +1,107 @@
+## Design Document - v0
+
+This is an oversimplified version of a kademlia/mainline DHT.
+
+For at least the v0 of our DHT we only care about storing and retrieving information about other DHT clients/servers; which we will call Peers.
+
+Mainline separates peers from Peers as one is the DHT client/server and the other is a BitTorrent provider.  
+We do not care about anything more than the DHT clients/servers for now, so this should simplify things a bit.
+
+## Terminology
+
+* `peer` - A participating client/server in the DHT network.
+* `k-buckets` - The routing table keeps a list of known GOOD peers.
+* `alpha` - A small number representing the degree of parallelism in network calls, usually 3.
+* `B` - The size in bits of the keys used to identify peers; in Kademlia this is 160, the length of a SHA1 digest.
+* `k` - The maximum number of contacts stored in a bucket; this is 20 in kademlia and 8 in mainline.
+* `tRefresh` - The time after which an otherwise unaccessed bucket must be refreshed; 3600s in Kademlia.
+* `refresh` - The process of getting fresh peers into unaccessed buckets.
+
+## Peer
+
+Each peer has an ID, for the purpose of this document we assume that the each ID is unique in the network.
+
+Kademlia uses 160-bit identifiers (SHA1) for peer IDs, we could explore the possibility of moving to some form of multihash at some point if we decide that this DHT does not have to be compatible with Kad.
+
+### XOR Metric & Distance Calculation
+
+Kademlia's operations are based upon the use of exclusive OR, XOR, as a metric. The distance between any two IDs x and y is defined as
+
+    distance(x, y) = x XOR y
+
+### The K-Bucket List (Routing Table)
+
+A Kademlia peer organizes its contacts (other peers known to it) in buckets which hold a maximum of k peers. These are known as k-buckets.
```

Comment:
Not sure why you'd need a binary tree for this. A map/slice would be probably enough?

eg. Assuming:

* `B=160`
* `k=20`

Your routing table would be as simple as `rt [160][20]*Contact`  
Finding your bucket is just a matter of hashing the id to find the index between 0-160.

An alternative to this would be using a [doubly linked list](https://golang.org/pkg/container/list/) for each bucket so you can easier move things to head/tail. eg. `[160]*List` with your `Peer.ID` being stored in the `Element.Value`.